Add experimental support for using Mojmap as a runtime

EnnuiL opened the pull request:
~~This expands on @TheGlitch76's initial efforts on running Mojmap as runtime; It works! The log messages are all translated to Mojmap instead of Intermediary! And in theory? Mods that only have code without any use of Mixins and AWs can work!~~

~~The issue is: the remapper is wonky! It currently hates Mixins and AWs *and* some results may be partial! So my experiment is y'all's problem now :p~~

~~(PS: The removal of Tiny v1 support will require Quilt Meta to provide Intermediary v2, which I have tested and can confirm it'll be fine)~~


^ Outdated -- see https://github.com/QuiltMC/quilt-loader/pull/469#issuecomment-2625752778 below
(glitchhhhhhhh I can't believe you have hijacked this PR into a successful thing!)

EnnuiL committed:
Begin progress on Mojmap as runtime

TheGlitch76 committed:
Clarify that the QuiltLauncher is the source of truth on the runtime namespace

TheGlitch76 committed:
Work for Oro

TheGlitch76 committed:
It works! Still needs lots of cleaning...

TheGlitch76 committed:
Get things in a working state

TheGlitch76 committed:
Clean up

TheGlitch76 committed:
Hide intermediate mappings once again

TheGlitch76 committed:
Final fixes(?)

TheGlitch76 committed:
Fix tests

TheGlitch76 committed:
License header

TheGlitch76 commented:
This PR is ready for review now. It allows selecting Mojmap as the runtime namespace in prod, and it uses RuntimeModRemapper to remap all mods in intermediary to mojmap (It successfully loaded the Cobblemon modpack, but might not be fully reliable). Without the experimental system properties, Loader should behave exactly as before.

Some more work will be needed before a hypothetical switch of defaults could occur, at minimum:
- Figuring out how Loader will actually find mojmaps (just download them itself?)
- Figuring out how Mojmap will be named in mod metadata (I'm actually not a fan of the maven identifier, since we're using something that's explicitly not on maven)
- Testing of loading mojmap-compiled mods in a prod mojmap environment

In the interest of getting this merged in a reasonable timeframe, the remaining work will be covered in a different PR.

OroArmor commented on src/main/java/org/quiltmc/loader/impl/metadata/qmj/V1ModMetadataReader.java:
```diff
@@ -285,7 +285,8 @@ private V1ModMetadataImpl readFields(JsonLoaderValue.ObjectImpl root) {
 
-			String[] supported_mappings = { "org.quiltmc:hashed", "net.fabricmc:intermediary" };
+			// TODO: "mojang" for mojmap breaks spec; it is just temporary for now because its experimental
+			String[] supported_mappings = { "org.quiltmc:hashed", "net.fabricmc:intermediary", "mojang" };
```
`"net.minecraft:mojang"`?

EnnuiL replied:
glitch says this is temporary, but well? I know how permanent placeholders can be like!
Anyway, `net.minecraft:mojang` makes sense for dev purposes, but Loom uses `net.minecraft:mappings`, which makes sense but is confusing! (what is Minecraft mappings? I've heard of Mojang!)

So, considering this and Mojang's official standards? Maybe we should go with `com.mojang:mappings`?

OroArmor commented on src/main/java/org/quiltmc/loader/impl/plugin/quilt/StandardQuiltPlugin.java:
```diff
@@ -394,9 +394,13 @@ private ModLoadOption[] scan0(Path root, QuiltLoaderIcon fileIcon, ModLocation l
-			boolean requiresRemap = !location.onClasspath() && QuiltLoader.isDevelopmentEnvironment();
+			// a mod needs to be remapped if the mod did not come from the classpath and its mappings do not match
+			// the current runtime namespace
+			String mappings = meta.intermediateMappings();
```
Id use a switch here

TheGlitch76 replied:
this is java 8

OroArmor replied:
you can switch on strings

TheGlitch76 committed:
Fix OOM in junit

EnnuiL commented on src/main/java/org/quiltmc/loader/impl/metadata/qmj/V1ModMetadataReader.java:
```diff
@@ -309,6 +310,10 @@ private V1ModMetadataImpl readFields(JsonLoaderValue.ObjectImpl root) {
 			}
 
+			if (mappings.equals("mojang") && !QuiltLoader.getMappingResolver().getNamespaces().contains("mojang")) {
+				throw new ParseException("Oh no! This version of Quilt Loader doesn't support Mojang mappings, please update Quilt Loader to use this mod.");
```
This copy-pastey comment doesn't make sense here compared to the Hashed Mojmap one; how about:
```suggestion
				throw new ParseException("Oh no! This Quilt Loader installation isn't configured to support Mojang mappings.");
```
(potential remedies to this are to be determined once uhhhh, we figure out a good way to provide mojmap)

EnnuiL reviewed (commented):
Something I do genuinely wonder about is: what about the mods that uses the remapper thingy? I wouldn't mind breaking compat, but I know that can be tricky; maybe #458 can help?
Anyway, here's the drill:

TheGlitch76 committed:
Use `org.quiltmc.experimental:mojang` as the identifier for mojmap for now

TheGlitch76 commented:
I've changed the `intermediate_mappings` field to be `org.quiltmc.experimental:mojang` which follows spec. Since this is currently in the same state of usability as plugins (actually, even less so!) I hope this can be an acceptable solution for now.

TheGlitch76 committed:
Move mapping management to the `GameProvider`

This includes (re-)adding support for
situations where there are no mappings loaded.

AlexIIL commented on src/main/java/org/quiltmc/loader/impl/game/MappingConfigurationImpl.java:
```diff
@@ -200,9 +204,34 @@ private void initialize() {
+				if (getTargetNamespace().equals("official")) {
+					Log.warn(LogCategory.MAPPINGS, "Continuing without mappings because target namespace is 'official'");
+				} else {
+					throw new IllegalStateException("Requested target namespace %s not loaded. To continue without mappings, " +
```
You either need to use string concatenation here or an explicit `String.format` to handle the `%s` as a target namespace
```suggestion
					throw new IllegalStateException("Requested target namespace " + targetNamespace + " not loaded. To continue without mappings, " +
```

AlexIIL commented on src/main/java/org/quiltmc/loader/impl/transformer/RuntimeModRemapper.java:
```diff
@@ -61,28 +69,59 @@ final class RuntimeModRemapper {
+			Log.info(LogCategory.MOD_REMAP, "Not remapping mods because mappings are empty");
+			return;
+		} else if (!mappingConfiguration.getNamespaces().contains("intermediary")) {
+			Log.warn(LogCategory.MOD_REMAP, "Not remapping mods because intermediary is missing!");
```
I think you're missing a `return` here, as this will crash if intermediary is actually missing.

```suggestion
			Log.warn(LogCategory.MOD_REMAP, "Not remapping mods because intermediary is missing!");
			return;
```

AlexIIL commented on src/main/java/org/quiltmc/loader/impl/transformer/KotlinMetadataRemapper.java:
```diff
@@ -0,0 +1,80 @@
+	@Override
+	public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
+		if (descriptor.equals("Lkotlin/Metadata;")) {
+			return new AnnotationNode(QuiltLoaderImpl.ASM_VERSION, descriptor) {
```
This doesn't send the modified data back to the parent ClassVisitor, you'd need to either add `
accept(KotlinMetadataRemapper.super.visitAnnotation(descriptor, visible));` to both of the returns in `visitEnd`, or replace this with an inline version.

*(I like the inline approach since it avoids building a partial node of a (potentially) long data structure, but just adding the `accept` version twice is the simplest change)*
```suggestion
			return new AnnotationVisitor(api, super.visitAnnotation(descriptor, visible)) {
				@Override
				public AnnotationVisitor visitArray(String name) {
					if ("d2".equals(name)) {
						return new AnnotationVisitor(api, super.visitArray(name)) {
							@Override
							public void visit(String name, Object value) {
								if (value instanceof String) {
									String candidate = (String) value;
									if (candidate.startsWith("(")) {
										candidate = remapper.mapMethodDesc(candidate);
									} else if (candidate.startsWith("L")) { // this could technically catch strays but it should just not do anything
										candidate = remapper.mapDesc(candidate);
									} else if (candidate.startsWith("class_") || candidate.contains("/")) { // must go last to not accidentally catch descriptors
										candidate = remapper.map(candidate);
									} // else hope nothing goes wrong
									value = candidate;
								}

								super.visit(name, value);
							}
						};
					}
					return super.visitArray(name);
				}
			};
		
```

AlexIIL reviewed (changes requested):
I've run this against my tests and found a few problems, but otherwise I like the direction!

AlexIIL commented:
I've pushed all my fixes here: https://github.com/EnnuiL/quilt-loader/pull/1 - feel free to either merge as-is (possibly manually @TheGlitch76 ) or tweak them further?

AlexIIL committed:
A few fixes:

- Fix MappingConfigurationImpl's "missing namespace" exception not including the target namespace in the error message
- Fix KotlinMetadataRemapper discarding the metadata annotation
- Slightly optimise and simplify the logic in KotlinMetadataRemapper by using AnnotationVisitor instead of AnnotationNode
- Fix a JVM hang when the remapper throws something that isn't an IOException after creating the TinyRemapper by wrapping it's entire lifetime in a try...finally { remapper.finish(); }
	- This ensures the non-daemon thread pool is stopped.
- Refactor how RuntimeModRemapper is used by TransformCache - the remapper is now an initialised object, and stores the "modsToRemap"
	- This allows the remapping logic to be fully stored in the remapper, which ensures the mod file copy step will always copy mod class files if the remapper isn't going to.
	  (Previously the cache assumed that if a mod had a "namespaceMappingFrom" then it would be remapped, however this is no longer the case with the checks at the start of RuntimeModRemapper)

TheGlitch76 committed:
Merge branch 'develop' into wip/mojmap-runtime

TheGlitch76 committed:
Remove outdated TODO

AlexIIL reviewed (approved):
Looks good, thanks!
